packages/queued-request-controller/src/QueuedRequestMiddleware.test.ts

```diff
@@ -1,607 +1,320 @@
-    jest.fn().mockReturnValue({
-      configuration: {
-        chainId: '0x1',
-      },
-    });
-  messenger.registerActionHandler(
-    'NetworkController:getNetworkClientById',
-    mockGetNetworkClientById,
-  );
-
-  const mockGetNetworkConfigurations =
-    mocks.getNetworkConfigurations ?? jest.fn(() => ({}));
-  const mockGetProviderConfig =
-    mocks.getProviderConfig ??
-    // TODO: Replace `any` with type
-    // eslint-disable-next-line @typescript-eslint/no-explicit-any
-    jest.fn<ProviderConfig, any[]>(() => ({
-      chainId: '0x1',
-      type: NetworkType.mainnet,
-      ticker: 'ETH',
-    }));
-  const mockGetNetworkControllerState = jest.fn(() => ({
-    ...networkControllerDefaultState,
-    networkConfigurations: mockGetNetworkConfigurations(),
-    providerConfig: mockGetProviderConfig(),
-  }));
-
-  messenger.registerActionHandler(
-    'NetworkController:getState',
-    mockGetNetworkControllerState,
-  );
-
-  const mockEnqueueRequest = jest.fn().mockImplementation((cb) => cb());
-  messenger.registerActionHandler(
-    'QueuedRequestController:enqueueRequest',
-    mockEnqueueRequest,
-  );
-
-  const mockAddRequest = mocks.addRequest ?? jest.fn().mockResolvedValue(true);
-  messenger.registerActionHandler(
-    'ApprovalController:addRequest',
-    mockAddRequest,
-  );
-
-  const mockSetActiveNetwork = jest.fn().mockResolvedValue(true);
-  messenger.registerActionHandler(
-    'NetworkController:setActiveNetwork',
-    mockSetActiveNetwork,
-  );
-
-  const mockSetNetworkClientIdForDomain = jest.fn().mockResolvedValue(true);
-  messenger.registerActionHandler(
-    SelectedNetworkControllerActionTypes.setNetworkClientIdForDomain,
-    mockSetNetworkClientIdForDomain,
-  );
+import { createQueuedRequestMiddleware } from './QueuedRequestMiddleware';
+import type { QueuedRequestMiddlewareJsonRpcRequest } from './types';
 
+const getRequestDefaults = (): QueuedRequestMiddlewareJsonRpcRequest => {
   return {
-    getProviderConfig: mockGetProviderConfig,
-    getNetworkConfigurations: mockGetNetworkConfigurations,
-    getNetworkControllerState: mockGetNetworkControllerState,
-    getNetworkClientById: mockGetNetworkClientById,
-    enqueueRequest: mockEnqueueRequest,
-    addRequest: mockAddRequest,
-    setActiveNetwork: mockSetActiveNetwork,
-    setNetworkClientIdForDomain: mockSetNetworkClientIdForDomain,
+    method: 'doesnt matter',
+    id: 'doesnt matter',
+    jsonrpc: '2.0' as const,
+    origin: 'example.com',
+    networkClientId: 'mainnet',
   };
 };
 
-const requestDefaults = {
-  method: 'doesnt matter',
-  id: 'doesnt matter',
-  jsonrpc: '2.0' as const,
-  origin: 'example.com',
-  networkClientId: 'mainnet',
+const getPendingResponseDefault = (): PendingJsonRpcResponse<Json> => {
+  return {
+    id: 'doesnt matter',
+    jsonrpc: '2.0' as const,
+  };
 };
 
+const getMockEnqueueRequest = (error?: unknown) =>
+  jest.fn().mockImplementation((_origin, requestNext) => requestNext(error));
```

Comment:
The `jest.fn` `any` type seems to be obscuring an error here:

1. If `enqueueRequest` return value is typed as `Promise<void>`,

```suggestion
  jest
    .fn<
      ReturnType<QueuedRequestController['enqueueRequest']>, // Promise<void>
      Parameters<QueuedRequestController['enqueueRequest']>
    >()
    .mockImplementation((_origin, requestNext) => requestNext(error));
```
we get this error in the same line
```ts
Argument of type '(_origin: QueuedRequestMiddlewareJsonRpcRequest, requestNext: (...arg: unknown[]) => Promise<unknown>) => Promise<unknown>' is not assignable to parameter of type '(request: QueuedRequestMiddlewareJsonRpcRequest, requestNext: (...arg: unknown[]) => Promise<unknown>) => Promise<void>'.
  Type 'Promise<unknown>' is not assignable to type 'Promise<void>'.
    Type 'unknown' is not assignable to type 'void'.ts(2345)
```

2. If `enqueueRequest` return value is typed as `Promise<unknown>`
```suggestion
  jest
    .fn<
      Promise<unknown>,
      Parameters<QueuedRequestController['enqueueRequest']>
    >()
    .mockImplementation((_origin, requestNext) => requestNext(error));
```

We get this error in every `enqueueRequest` param of all `createQueuedRequestMiddleware` calls in this test file.

```ts
Type 'Mock<Promise<unknown>, [request: QueuedRequestMiddlewareJsonRpcRequest, requestNext: (...arg: unknown[]) => Promise<unknown>]>' is not assignable to type '(request: QueuedRequestMiddlewareJsonRpcRequest, requestNext: (...arg: unknown[]) => Promise<unknown>) => Promise<void>'.
  Type 'Promise<unknown>' is not assignable to type 'Promise<void>'.
    Type 'unknown' is not assignable to type 'void'.ts(2322)
    
QueuedRequestMiddleware.ts(65, 3): The expected type comes from property 'enqueueRequest' which is declared here on type '{ enqueueRequest: (request: QueuedRequestMiddlewareJsonRpcRequest, requestNext: (...arg: unknown[]) => Promise<unknown>) => Promise<void>; useRequestQueue: () => boolean; }'
```

Reply:
This seems to work as a fix:

```diff
diff --git a/packages/queued-request-controller/src/QueuedRequestController.ts b/packages/queued-request-controller/src/QueuedRequestController.ts
index eea0b1d12..1760f8b80 100644
--- a/packages/queued-request-controller/src/QueuedRequestController.ts
+++ b/packages/queued-request-controller/src/QueuedRequestController.ts
@@ -223,7 +223,7 @@ export class QueuedRequestController extends BaseController<
   async enqueueRequest(
     request: QueuedRequestMiddlewareJsonRpcRequest,
     requestNext: (...arg: unknown[]) => Promise<unknown>,
-  ) {
+  ): ReturnType<typeof requestNext> {
     this.#updateCount(1);
     if (this.state.queuedRequestCount > 1) {
       try {
@@ -234,6 +234,7 @@ export class QueuedRequestController extends BaseController<
       }
     }
 
+    let promiseToReturn: Awaited<ReturnType<typeof requestNext>>;
     const processCurrentRequest = async () => {
       try {
         if (
@@ -242,8 +243,7 @@ export class QueuedRequestController extends BaseController<
         ) {
           await this.#switchNetworkIfNecessary(request);
         }
-
-        await requestNext();
+        promiseToReturn = await requestNext();
       } finally {
         // The count is updated as part of the request processing to ensure
         // that it has been updated before the next request is run.
@@ -253,5 +253,6 @@ export class QueuedRequestController extends BaseController<
 
     this.currentRequest = processCurrentRequest();
     await this.currentRequest;
+    return promiseToReturn;
   }
 }
```

Reply:
I think 80f305c29084a9bead31942cbd3122fcd34409a4 resolves this concern, let me know if not though.

Re: the `jest.fn` typing, do you think we should explicitly provide type parameters everywhere `jest.fn` is used? In the example you shared in a comment, it seemed very verbose. I am hesitant about that as a guideline unless we can find a way to do that more concisely.

Reply:
Since we are using `mockImplementation` here, is Jest able to infer the type of the function for us if we use the `jest.fn(() => { ... })` form?

Or is this something that upgrading to a newer version of Jest would help us with, since the TypeScript support seems to be more robust?

Reply:
@Gudahtt The simplification seems to cause a new error:

```ts
Expected 0 arguments, but got 1.ts(2554)
(parameter) error: unknown
```

The error won't show up unless the generic arguments are supplied (and the `any` types removed).

```suggestion
  jest
    .fn<
      ReturnType<QueuedRequestControllerEnqueueRequestAction['handler']>,
      Parameters<QueuedRequestControllerEnqueueRequestAction['handler']>
    >()
    .mockImplementation((_origin, requestNext) => requestNext(error));
```

Reverting or updating `requestNext` to one of the following fixes the issue.

- line 225 of `QueuedRequestController.ts`:
```ts
requestNext: (...args: unknown[]) => Promise<void>
requestNext: (error?: unknown) => Promise<void>
```

Reply:
@mcmire Our current version of jest doesn't infer based on the `mockImplementation` callback's type, but instead defaults to `any` if generic arguments are omitted i.e. if both are omitted we get `jest.Mock<any, any>`. Every time I've come across this, including here, it's been a dangerous silent failure. I think we should err on the side of caution on this and remove the `any` typing wherever we can, since it specifically suppresses the basic functionality of tests: alerting us if anything goes wrong.

@Gudahtt Good news regarding verbosity is that v29 `jest.Mock<T>` only takes one function-type generic parameter, instead of return type and parameters type separately. That should make it less cumbersome to supply the generic argument and remove the `any`: 

```ts
  jest
    .fn<QueuedRequestController['enqueueRequest']>()
    .mockImplementation((_origin, requestNext) => requestNext(error));
```

As for inference, based on the [jest docs](https://jestjs.io/docs/mock-function-api#jestfnimplementation), it seems that in v29 `mockImplementation` infers the type of its callback's arguments and return value based on the generic argument `T` of `jest.fn<T>`, rather than the other way around, meaning supplying that generic argument while writing tests will still be important.

Reply:
Ah sorry, I forgot to update this line to remove the error parameter :facepalm: 
Fixed in f71ed230817e49864a74a7baa59b74ee8f5f81ce

Reply:
@MajorLift A single type parameter does look much easier to work with, that's good to hear.

Maybe we can make a greater effort to ensure we fill in type parameters once we update to Jest v29? Updating first might save us some time.

Reply:
Sounds like a plan!